File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/publish/DataTrackEmitter.kt
package com.swmansion.moqkit.publish

import android.os.SystemClock
import dev.moq.Frame
import dev.moq.TrackProducer

/**
 * Sends raw binary payloads on a data track.
 *
 * Create an emitter, pass it to [Publisher.addDataTrack], publish and start the
 * [Publisher], then call [send] whenever the app has a payload to deliver.
 */
class DataTrackEmitter {
    @Volatile private var producer: TrackProducer? = null
    @Volatile private var clock: Clock? = null
    @Volatile private var stopped = false

    internal fun attach(producer: TrackProducer, clock: Clock) {
        stopped = false
        this.clock = clock
        this.producer = producer
    }

    internal fun detach() {
        stopped = true
        producer = null
        clock = null
    }

    /**
     * Sends one binary payload.
     *
     * If the publisher has not started yet, or if the data track has already stopped, this
     * call is ignored.
     */
    fun send(data: ByteArray) {
        if (stopped) return
        val producer = producer ?: return
        // Stamp against the publisher clock so data frames share the epoch of the
        // broadcast's media tracks.
        val nowUs = SystemClock.elapsedRealtimeNanos() / 1_000
        val timestampUs = clock?.timestampUs(nowUs) ?: 0L
        producer.writeFrame(Frame(payload = data, timestampUs = timestampUs.toULong()))
    }
}

File: ios/Sources/MoQKit/Publish/DataTrackEmitter.swift
import CoreMedia
import Foundation
import MoqFFI

/// Push-based source for publishing app-defined binary messages on an object track.
///
/// Create one emitter per published data track, hand it to ``Publisher/addDataTrack(name:source:)``,
/// then keep a reference and call ``send(_:)`` whenever your app has a new payload.
public final class DataTrackEmitter: @unchecked Sendable {
    private var producer: MoqTrackProducer?
    private var clock: PublisherClock?
    private var stopped = false

    /// Creates an emitter that can be attached to a published data track.
    public init() {}

    internal func attach(_ producer: MoqTrackProducer, clock: PublisherClock) {
        self.clock = clock
        self.producer = producer
    }

    internal func detach() {
        stopped = true
        producer = nil
        clock = nil
    }

    /// Publishes one object on the track.
    ///
    /// If the track has not started yet, or has already stopped, this is a no-op.
    public func send(_ data: Data) throws {
        guard !stopped, let producer else { return }
        // Stamp against the publisher clock so data frames share the epoch of the
        // broadcast's media tracks.
        let now = CMClockGetTime(CMClockGetHostTimeClock())
        let timestampUs = clock?.timestampUs(from: now) ?? 0
        try producer.writeFrame(frame: MoqFrame(payload: data, timestampUs: timestampUs))
    }
}
